botorch/acquisition/scorebo.py

```diff
@@ -0,0 +1,186 @@
+        # to get num_optima x num_gps unique GPs
+        self.maximize = maximize
+        if not self.maximize:
+            optimal_outputs = -optimal_outputs
+
+        # inputs come as num_optima_per_model x num_models x d
+        # but we want it four-dimensional to condition one per model.
+
+        self.optimal_outputs = optimal_outputs.unsqueeze(-2)
+        # JES-like version of SCoreBO if optimal inputs are provided
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore")
+            with fantasize_flag():
+                with settings.propagate_grads(False):
+                    # We must do a forward pass one before conditioning.
+                    self.model.posterior(
+                        self.model.train_inputs[0], observation_noise=False
+                    )
+
+        if optimal_inputs is not None:
+            self.optimal_inputs = optimal_inputs.unsqueeze(-2)
+            self.conditional_model = self.model.condition_on_observations(
+                X=self.model.transform_inputs(self.optimal_inputs),
+                Y=self.optimal_outputs,
+                noise=torch.full_like(self.optimal_outputs, MIN_INFERRED_NOISE_LEVEL),
+            )
+
+        # otherwise, we do a MES-like variant (which places vastly more emphasis on
+        # HP learning as supposed to optimization)
+        else:
+            self.conditional_model = self.model
+
+        # Currently only supports LB (lower bound) estimation, will add MC later on
+        if estimation_type not in ["LB"]:
+            raise ValueError(f"Estimation type {estimation_type} does not exist.")
+        self.estimation_type = estimation_type
+
+        # the default number of MC samples (512) are too many when doing FB modeling.
+        if distance_metric not in DISTANCE_METRICS.keys():
+            raise ValueError(
+                f"Distance metric need to be one of {list(DISTANCE_METRICS.keys())}"
+            )
+        self.distance = DISTANCE_METRICS[distance_metric]
+        self.set_X_pending(X_pending)
+
+    @concatenate_pending_points
+    @t_batch_mode_transform()
+    def forward(self, X: Tensor) -> Tensor:
+        # since we have two MC dims (over models and optima), we need to
+        # unsqueeze a second dim to accomodate the posterior pass
+        prev_posterior = self.model.posterior(
+            X.unsqueeze(MCMC_DIM), observation_noise=True
+        )
+        posterior = self.conditional_model.posterior(
+            X.unsqueeze(MCMC_DIM), observation_noise=True
+        )
+        cond_means = posterior.mean
+        marg_mean = prev_posterior.mean.mean(dim=MCMC_DIM, keepdim=True)
+        cond_variances = posterior.variance
+        cond_covar = posterior.covariance_matrix
+
+        # the mixture variance is squeezed, need it unsqueezed
+        marg_covar = prev_posterior.mixture_covariance_matrix.unsqueeze(MCMC_DIM)
+        noiseless_var = self.conditional_model.posterior(
```

Comment:
if you need both noisy and noiseless variance, it would be more efficient to get the noiseless posterior and pass it through the likelihood manually to get the noisy one.

Reply:
Modified!

Reply:
It did squeeze the trailing dimension though, so I had to do some unsqueezing. Is this intended?